File: exercism.io/kotlin/beer-song/src/main/kotlin/BeerSong.kt
object BeerSong {
    val idxList: List<String> = listOf("no more bottles", "1 bottle").union((2..99).map { it.toString() + " bottles" }.toSet()).toList()

    fun verse(idx: Int) : String = when {
        idx < 0 || idx > 99 -> throw IllegalArgumentException()
        else -> "${idxList[idx].capitalize()} of beer on the wall, ${idxList[idx]} of beer.\n${part3(idx)} ${idxList[(idx + 99) % 100]} of beer on the wall.\n"
    }

    private fun part3(i: Int) : String = when (i) {
        0 -> "Go to the store and buy some more,"
        else -> "Take ${if (i == 1) "it" else "one"} down and pass it around,"
    }

    fun verses(a: Int, b: Int) : String = when {
        a > 99 || b > 99 || a < 0 || b < 0 -> throw IllegalArgumentException()
        else -> (Math.min(a, b)..Math.max(a, b)).map { verse(it) }.reversed().joinToString(separator="\n")
    }

    val lyrics = verses(0, 99) // both (0, 99) and (99, 0) work!! :P
}

File: exercism.io/javascript/beer-song/beer-song.js
/**
 * Created by mewx on 19/12/16.
 */

var BeerSong = function () {
    this.verse = function (a) {
        return this.cap(this.head(a)) +  " of beer on the wall, " + this.head(a) + " of beer.\n" +
            this.part3(a) + " " + this.head((a + 99) % 100) + " of beer on the wall.\n";
    };

    this.sing = function (a, b = 0) {
        var result = this.verse(a --);
        while (a >= b) {
            result = result.concat("\n" + this.verse(a));
            a -= 1;
        }
        return result;
    };

    this.cap = function (str) {
        return str[0].toUpperCase() + str.substring(1);
    };

    this.head = function (i) {
        switch (i) {
            case 0:
                return "no more bottles";
            case 1:
                return "1 bottle";
            default:
                return i + " bottles";
        }
    };

    this.part3 = function (i) {
        var it;
        if (i == 1) it = "it";
        else it = "one";

        if (i == 0) return "Go to the store and buy some more,";
        else return "Take " + it + " down and pass it around,";
    };
};

module.exports = BeerSong;
